### DISCUSSION TITLE
Link Prediction on Heterogeneous Graphs/ Recommendation system

### DISCUSSION THREAD

**Discussion by radojicicmarinaa** [pyg-team/pytorch_geometric#4409] [Q&A]
Hello, I have been working on a project that involves link prediction using Heterogeneous graph. ( data are read from neo4j and are then elaborated using PyTorch geometric)
The idea of the project can be represented in the same way as in the article about Movies recommendations. 
( Movie recommendation tutorial I followed: https://github.com/tomasonjo/blogs/blob/master/pyg2neo/Movie_recommendations.ipynb )

I have 3 nodes: Movie, Cinema and User. 
I cannot consider all the movies to recommend, as a user can only watch movies in a specific cinema and cinemas between them have different film timetables. ( for the sake of simplicity we will consider that timetables and list of films do not change during the time)
Goal: I would like to predict the movie the user will see next. 

In the graph DB two nodes are connected IFF a user saw a specific film. ( edge User - Movie)
One of the things that I noticed is that when I am trying to apply the Heterodata class and see the obtained edge labels, I only get 1 class. 
( code example: data['user', 'watch', 'asset'].edge_label = edge_label ) 
On the other hand, in the Movie recommendation article there were 6 classes: from 0 to 5. ( multi class problem) 

In my case, edge label contains only label that is equal to 1 and edge_label has the same length as the edge_index. 

Could you please help me understand whether this behaviour is normal, or I should do some changes in the DB in order to obtain both 0 and 1 labels after applying HeteroData(). ( 0 = the link does not exist, 1 = the link exists) 

Last but not least, could this thing be handled in the  RandomLinkSplit function adding parameter neg_sampling_ratio ? 

Thanks in advance.

**Comment by rusty1s** [comment]
Yes, this is normal. In the MovieLens example, we are actually doing link-level classification, and we do not infer missing links. You can easily enfore that by setting `neg_sampling_ratio` in `RandomLinkSplit` to a value greater than 0 (for example, 1 or 2). This way, `edge_label` will contain both `0` and `1` entries, and negative edges (links that do not exist in your graph) will get sampled.

**Reply by radojicicmarinaa** [reply]
Hello @rusty1s , thanks for your reply!

Actually today I tried adding the parameter and I got 2 classes instead of one!
However, I have still one doubt left. If we consider a hypothetical relationship between User and Cinema like : User WORKS_AT Cinema and a user can just see JUST the movies from the cinema he works at, how can I know that I will recommend the user just the movies from his cinema?   
Can I be sure that using the embeddings ( like fastrp or graphsage - where we take the information about the neighbour nodes ) that to a specific user will be suggested just the films from the cinema he works at? 
Is there any specific way to deal this kind of conditions? 
 
Thanks!:)

**Reply by rusty1s** [reply]
I think this is highly related to logical reasoning in knowledge graphs, see https://arxiv.org/abs/2002.05969. In your case, if you already know that a user works at a certain cinema, you can simply before ranking on the movies from this particular cinema. Would that work in your scenario?
